router/batchrouter/asyncdestinationmanager/bing-ads/offline-conversions/bingads_test.go

```diff
@@ -0,0 +1,653 @@
+	"github.com/rudderlabs/rudder-go-kit/config"
+	"github.com/rudderlabs/rudder-go-kit/logger"
+	backendconfig "github.com/rudderlabs/rudder-server/backend-config"
+	"github.com/rudderlabs/rudder-server/jobsdb"
+	mock_bulkservice "github.com/rudderlabs/rudder-server/mocks/router/bingads"
+	mocks_oauth "github.com/rudderlabs/rudder-server/mocks/services/oauth"
+	"github.com/rudderlabs/rudder-server/router/batchrouter/asyncdestinationmanager/common"
+	oauth "github.com/rudderlabs/rudder-server/services/oauth"
+	"github.com/rudderlabs/rudder-server/utils/misc"
+)
+
+var (
+	once        sync.Once
+	destination = backendconfig.DestinationT{
+		Name: "BingAds",
+		Config: map[string]interface{}{
+			"customerAccountId": "customer_account_id",
+			"customerId":        "customer_id",
+			"rudderAccountId":   "rudder_account_id",
+		},
+		WorkspaceID: "workspace_id",
+	}
+)
+
+var currentDir, _ = os.Getwd()
+
+func initBingads() {
+	once.Do(func() {
+		logger.Reset()
+		misc.Init()
+	})
+}
+
+var _ = Describe("Bing ads Offline Conversions", func() {
+	Context("Bing ads", func() {
+		BeforeEach(func() {
+			config.Reset()
+			config.Set("BatchRouter.BING_ADS_OFFLINE_CONVERSIONS.MaxUploadLimit", 1*bytesize.KB)
+		})
+
+		AfterEach(func() {
+			config.Reset()
+		})
+
+		It("TestBingAdsUploadPartialSuccessCase", func() {
+			initBingads()
+			ctrl := gomock.NewController(GinkgoT())
+			bingAdsService := mock_bulkservice.NewMockBulkServiceI(ctrl)
+			clientI := Client{
+				URL:    "http://localhost/upload1",
+				client: &http.Client{},
+			}
+			bulkUploader := NewBingAdsBulkUploader("BING_ADS", bingAdsService, &clientI)
+			bingAdsService.EXPECT().GetBulkUploadUrl().Return(&bingads_sdk.GetBulkUploadUrlResponse{
+				UploadUrl: "http://localhost/upload1",
+				RequestId: misc.FastUUID().URN(),
+			}, nil)
+			bingAdsService.EXPECT().GetBulkUploadUrl().Return(&bingads_sdk.GetBulkUploadUrlResponse{
+				UploadUrl: "http://localhost/upload2",
+				RequestId: misc.FastUUID().URN(),
+			}, nil)
+			bingAdsService.EXPECT().GetBulkUploadUrl().Return(&bingads_sdk.GetBulkUploadUrlResponse{
+				UploadUrl: "http://localhost/upload3",
+				RequestId: misc.FastUUID().URN(),
+			}, nil)
+
+			bingAdsService.EXPECT().UploadBulkFile("http://localhost/upload1", gomock.Any()).Return(&bingads_sdk.UploadBulkFileResponse{
+				TrackingId: "randomTrackingId1",
+				RequestId:  "randomRequestId1",
+			}, nil)
+			bingAdsService.EXPECT().UploadBulkFile("http://localhost/upload2", gomock.Any()).Return(nil, fmt.Errorf("unable to get bulk upload url, check your credentials"))
+			bingAdsService.EXPECT().UploadBulkFile("http://localhost/upload3", gomock.Any()).Return(&bingads_sdk.UploadBulkFileResponse{
+				TrackingId: "randomTrackingId3",
+				RequestId:  "randomRequestId3",
+			}, nil)
+			dir, err := os.MkdirTemp("/tmp", "rudder-server")
+			if err != nil {
+				fmt.Printf("Failed to create temporary directory: %v\n", err)
+				return
+			}
```

Comment:
If any of the `if err != nil` fails it would be hard to debug, since you print and just return, the test suite would not register any failure and silently return. Without verbose mode, logs are not going be shown as well.

Since you are using gomega, the correct way to handle errors is:

```suggestion
			Exepect(err).ShouldNot(HaveOccurred(), "creating temporary directory")
```

https://onsi.github.io/gomega/#haveoccurred

Similar of all the `if err != nil` in the code.

Reply:
Changed all err handling in test cases with `expect(err).ShouldNot(HaveOccurred(),'some_text')`